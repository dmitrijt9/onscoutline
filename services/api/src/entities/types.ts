/**
 * Common date format with time part - https://cs.wikipedia.org/wiki/ISO_8601
 * @example 2022-02-24T16:00:00.000Z
 */
export type ISO8601 = string

/**
 * Common date format without time part - https://cs.wikipedia.org/wiki/ISO_8601
 * @example 2022-02-24
 */
export type ISO8601_NoTime = string
