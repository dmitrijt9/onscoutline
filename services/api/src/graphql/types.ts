import { Container } from '../dependency/container/index'
import { ExpressContext } from 'apollo-server-express'

export type GraphqlContext = {
    container: Container
    // user: User
    // loaders: ContextLoaders
    express?: ExpressContext
}
