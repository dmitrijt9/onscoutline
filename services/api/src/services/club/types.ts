export type NewClubRequest = {
    facrId: string
    name: string
    facrUuid: string
}
