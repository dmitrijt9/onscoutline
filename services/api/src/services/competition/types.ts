export type NewCompetitionRequest = {
    regionName: string
    regionId: string
    facrId: string
    name: string
    facrUuid: string
}
