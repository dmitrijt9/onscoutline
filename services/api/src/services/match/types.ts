export type NewMatchRequest = {
    competition: string
    takePlace: string
    facrUuid: string
    homeTeam: string
    awayTeam: string
    homeTeamScore: number | null
    awayTeamScore: number | null
    lineups: LineupsRequest
}

type GoalRequest = {
    minute: number
    type: string
}

type LineupsRequest = {
    home: MatchPlayerRequest[]
    away: MatchPlayerRequest[]
}

type MatchPlayerRequest = {
    shirt: number
    position: string
    fullname: string
    yellowCard: boolean
    redCard: boolean
    substitution: string
    isInStartingLineup: boolean
    goals: GoalRequest[]
}
