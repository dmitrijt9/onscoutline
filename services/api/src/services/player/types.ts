import { Club } from '../../entities/Club'
import { Gender, Player, Transfer } from '../../entities/Player'
import { PlayerInClub } from '../../entities/Relations/PlayerInClub'
import { ISO8601_NoTime } from '../../entities/types'

export type NewPlayerRequest = {
    facrId: string

    name: string

    surname: string

    dateOfBirth: string

    facrMemberFrom?: ISO8601_NoTime

    gender: Gender

    country: string

    parentClub: {
        clubFacrId: string
        clubName: string
        playingFrom: ISO8601_NoTime
    }

    loanClub: {
        clubFacrId: string
        clubName: string
        playingFrom: ISO8601_NoTime
        playingUntil: ISO8601_NoTime
    } | null

    transfersRecords: Transfer[]
}

export type PlayerToUpdate = Player

export type PlayerInClubToSave = Omit<PlayerInClub, 'id'>

export type PlayerInClubRequest = Player & {
    parentClub: NewPlayerRequest['parentClub']
    loanClub: NewPlayerRequest['loanClub']
}
