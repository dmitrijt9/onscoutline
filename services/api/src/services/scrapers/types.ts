import { HTMLElement } from 'node-html-parser'
import { Club } from '../../entities/Club'
import { Competition } from '../../entities/Competition'
import { NewClubRequest } from '../club/types'
import { NewCompetitionRequest } from '../competition/types'
import { NewMatchRequest } from '../match/types'
import { NewPlayerRequest } from '../player/types'

export interface IScraper {
    getParsedPage(url: string): Promise<HTMLElement>
}

export interface IFacrScraper {
    scrapeCompetitions(): Promise<NewCompetitionRequest[]>
    scrapeClubs(dirname: string): Promise<NewClubRequest[]>
    scrapePlayersOfClubs(clubs: Club[]): Promise<Map<string, NewPlayerRequest[]>>
    scrapeMatches(htmlsToScrape: string[]): Promise<NewMatchRequest[]>
}

export type ScrapedClub = Omit<Club, 'id' | 'facrId' | 'facrUuid'> & {
    facrId: string
    facrUuid: string
}

export type ScrapedCompetition = Omit<Competition, 'id'>

export type TPuppeteerSelector = string | string[]

export type TPuppeteerSelectorOptions = {
    visible?: boolean
    hidden?: boolean
    timeout?: number
}

export type ScrapedMatchOverview = {
    facrUuid: NewMatchRequest['facrUuid']
    takePlace: NewMatchRequest['takePlace']
}
