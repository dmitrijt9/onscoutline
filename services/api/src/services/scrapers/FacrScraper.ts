import { HTMLElement } from 'node-html-parser'
import { AppConfig } from '../../dependency/config/index'
import { CompetitionRepository } from '../../repositories/CompetitionRepository'
import { IScraper } from './types'

export class FacrScraper {
    private facrCompetitionsUrl: string
    // private facrMembersUrl: string

    // format of the competitions path is -> `${COMPETITION_X_PAGE_PATH_PREFIX}/[UUID]`
    // private COMPETITION_MAIN_PAGE_PATH_PREFIX = '/turnaje/hlavni'
    private static readonly COMPETITION_CLUBS_PAGE_PATH_PREFIX = '/turnaje/team'
    // private COMPETITION_MATCHES_PAGE_PATH_PREFIX = '/turnaje/zapas'

    constructor(
        private readonly scraper: IScraper,
        { facrScraper }: AppConfig,
        private readonly competitionRepository: CompetitionRepository,
    ) {
        this.facrCompetitionsUrl = facrScraper.facrCompetitionsUrl
        // this.facrMembersUrl = facrScraper.facrMembersUrl
    }

    // TODO: Save region paths to database
    async scrapeAndSaveCompetitions() {
        console.log('FACR Scraper: Starting to scrape competitions data.')

        try {
            const regionsParsedPage = await this.scraper.getParsedPage(
                `${this.facrCompetitionsUrl}/subjekty`,
            )

            const regionsPaths = regionsParsedPage.querySelectorAll('div.box a.btn').map((atag) => {
                return atag.getAttribute('href')
            })
            const competitionsBasicDataFetchers = regionsPaths.map(async (regionPath) => {
                // #souteze anchor ensures that page opens with opened tab that we need, where table with competitions is hidden
                const parsedCompetitionPage = await this.scraper.getParsedPage(
                    `${this.facrCompetitionsUrl}${regionPath}#souteze`,
                )
                const competitionRegionName =
                    parsedCompetitionPage.querySelector('h1.h1')?.innerText
                return parsedCompetitionPage
                    .querySelectorAll('#souteze table.table tbody tr')
                    .map((row: HTMLElement) => {
                        return {
                            regionName: competitionRegionName,
                            facrId: row.querySelector('td:nth-child(1)')?.innerText,
                            name: row.querySelector('td:nth-child(2) a')?.innerText,
                            facrUuid: row
                                .querySelector('td:nth-child(2) a')
                                ?.getAttribute('href')
                                ?.split('/')[3],
                        }
                    })
            })

            const competitionsData = await (
                await Promise.all(competitionsBasicDataFetchers).finally(() => {
                    console.log('FACR Scraper: Successfully scraped competitions data.')
                })
            ).flat()

            await this.competitionRepository.save(competitionsData)
        } catch (e) {
            console.error('FACR Scraper: Error while scraping competitions.', e)
        }
    }

    // TODO: save clubs to database
    async getScrapedClubs() {
        console.log('FACR Scraper: Starting to scrape clubs data.')

        try {
            const competitions = await this.competitionRepository.find()
            if (competitions.length <= 0) {
                console.log('FACR Scraper: No competitions to scrape clubs.')
                return
            }
            const clubsDataFetchers = competitions.map(async (competition) => {
                const clubsParsedPage = await this.scraper.getParsedPage(
                    `${FacrScraper.COMPETITION_CLUBS_PAGE_PATH_PREFIX}/${competition.facrUuid}`,
                )

                return clubsParsedPage
                    .querySelectorAll('.table-container table.table tbody tr')
                    .map((clubRow) => {
                        const nameColumn = clubRow.querySelector('td:nth-child(2)')
                        const clubLink = clubRow.querySelector('td:nth-child(2) a')
                        const idColumn = clubRow.querySelector('td:nth-child(3)')
                        if (!nameColumn) {
                            // TODO: custom error for scraper missconfigurations, clubRow as a payload
                            throw new Error(
                                `FACR Scraper: Failed to scrape clubs. Query selector: ${nameColumn}`,
                            )
                        }

                        if (!clubLink) {
                            // TODO: custom error for scraper missconfigurations, clubRow as a payload
                            throw new Error(
                                `FACR Scraper: Failed to scrape clubs. Query selector: ${clubLink}`,
                            )
                        }

                        if (!idColumn) {
                            // TODO: custom error for scraper missconfigurations, clubRow as a payload
                            throw new Error(
                                `FACR Scraper: Failed to scrape clubs. Query selector: ${idColumn}`,
                            )
                        }
                        return {
                            clubName: nameColumn.innerText,
                            clubFacrPath: clubLink.getAttribute('href'),
                            clubFacrId: idColumn.innerText,
                            clubFacrUuid: clubLink.getAttribute('href')?.split('/')[3],
                        }
                    })
                    .flat()
            })

            const clubs = (
                await Promise.all(clubsDataFetchers).finally(() => {
                    console.log('FACR Scraper: Successfully scraped clubs data.')
                })
            ).flat()

            // To ensure that there are no duplicates
            const clubsSet = new Set(clubs)

            // Convert Set back to array
            return [...clubsSet]
        } catch (e) {
            console.error('FACR Scraper: Error while scraping clubs.', e)
        }
    }
}
