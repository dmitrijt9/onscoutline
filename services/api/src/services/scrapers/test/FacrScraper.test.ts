import { readFileSync } from 'fs'
import parse from 'node-html-parser'
import {
    cleanDb,
    createTestingClient,
    stopTestApplication,
    TestingClient,
} from '../../../dependency/test-utils/index'
import { FacrScraper } from '../FacrScraper'
import { IFacrScraper } from '../types'

jest.setTimeout(100 * 1000)

describe('FacrScraper', () => {
    let testingClient: TestingClient
    let facrScraper: IFacrScraper

    beforeAll(async () => {
        testingClient = await createTestingClient()
        facrScraper = testingClient.container.facrScraper
    })

    afterAll(async () => {
        await stopTestApplication(testingClient.application)
    })

    afterEach(async () => {
        await cleanDb(testingClient)
    })

    it('should test the testing clinet', async () => {
        const regionsTestHtmlString = readFileSync(__dirname + '/mocks/facr-regions-test.html', {
            encoding: 'utf8',
            flag: 'r',
        })

        const mainCompetitionsTestHtmlString = readFileSync(
            __dirname + '/mocks/facr-main-competitions-test.html',
            {
                encoding: 'utf8',
                flag: 'r',
            },
        )
        const getParsedPageMock = jest
            .spyOn(FacrScraper.prototype, 'getParsedPage')
            .mockImplementation(async (url: string) => {
                if (url.endsWith('/subjekty')) {
                    return parse(regionsTestHtmlString)
                }

                return parse(mainCompetitionsTestHtmlString)
            })

        const scrapedCompetitions = await facrScraper.scrapeCompetitions()
        const savedCompetitions =
            await testingClient.container.competitionService.saveNewCompetitions(
                scrapedCompetitions,
            )

        expect(getParsedPageMock).toHaveBeenCalled()
        expect(savedCompetitions).not.toBeUndefined()
        expect(savedCompetitions?.length).toBe(2)
    })
})
