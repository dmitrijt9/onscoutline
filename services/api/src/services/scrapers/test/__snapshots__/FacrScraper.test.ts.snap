// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacrScraper should scrape matches from html correctly 1`] = `
Object {
  "awayTeam": "FK Teplice - fotbal, spolek",
  "awayTeamScore": 4,
  "competition": "Starší přípravka A",
  "facrUuid": "852d404e-687a-4595-867b-ea64b018dc78",
  "homeTeam": "VIKTORIA PLZEŇ - fotbal, z.s.",
  "homeTeamScore": 7,
  "lineups": Object {
    "away": Array [
      Object {
        "fullname": "Zulegerová Anna",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Brankář",
        "redCard": false,
        "shirt": 30,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Lišková Barbora",
        "goals": Array [
          Object {
            "minute": 9,
            "type": "Branka",
          },
          Object {
            "minute": 17,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 17,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Milíčková Michaela",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 18,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Redlichová Zdenička",
        "goals": Array [
          Object {
            "minute": 26,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 2,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Šandová Marie",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 22,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Špaková Nikola",
        "goals": Array [
          Object {
            "minute": 20,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 12,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Novotná Ema",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 10,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Vacínová Aneta Klára",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 11,
        "substitution": "",
        "yellowCard": false,
      },
    ],
    "home": Array [
      Object {
        "fullname": "Dragounová Barbora",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Brankář",
        "redCard": false,
        "shirt": 32,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Beranová Sára",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 8,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Šandová Sofie",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 12,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Kováříková Ema",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 15,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Pohořalá Monika",
        "goals": Array [
          Object {
            "minute": 4,
            "type": "Branka",
          },
          Object {
            "minute": 14,
            "type": "Branka",
          },
          Object {
            "minute": 17,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 18,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Štrosová Natálie",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 19,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Dragounová Adéla",
        "goals": Array [
          Object {
            "minute": 7,
            "type": "Branka",
          },
          Object {
            "minute": 21,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 25,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Valtrová Elena",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 6,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Procházková Eliška",
        "goals": Array [],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 20,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Vacíková Nela",
        "goals": Array [
          Object {
            "minute": 12,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 17,
        "substitution": "",
        "yellowCard": false,
      },
      Object {
        "fullname": "Marešová Petra",
        "goals": Array [
          Object {
            "minute": 23,
            "type": "Branka",
          },
        ],
        "isInStartingLineup": true,
        "position": "Neuvedeno",
        "redCard": false,
        "shirt": 21,
        "substitution": "",
        "yellowCard": false,
      },
    ],
  },
  "takePlace": "Neděle 26. 9. 2021 13:00",
}
`;
